src/common/resource.cu

```diff
@@ -0,0 +1,43 @@
+/**
+ * Copyright 2024, XGBoost Contributors
+ */
+#include "device_helpers.cuh"  // for CurrentDevice
+#include "resource.cuh"
+#include "xgboost/string_view.h"  // for StringView
+
+namespace xgboost::common {
+CudaMmapResource::CudaMmapResource(StringView path, std::size_t offset, std::size_t length)
+    : ResourceHandler{kCudaMmap},
+      handle_{detail::OpenMmap(std::string{path}, offset, length),
+              [](MMAPFile* handle) {
+                // Don't close the mmap while CUDA kernel is running.
+                if (handle) {
+                  dh::DefaultStream().Sync();
+                }
+                detail::CloseMmap(handle);
+              }},
+      n_{length} {
+  auto device = dh::CurrentDevice();
+  dh::safe_cuda(
+      cudaMemAdvise(handle_->base_ptr, handle_->base_size, cudaMemAdviseSetReadMostly, device));
+  dh::safe_cuda(cudaMemAdvise(handle_->base_ptr, handle_->base_size,
+                              cudaMemAdviseSetPreferredLocation, device));
+  dh::safe_cuda(
+      cudaMemAdvise(handle_->base_ptr, handle_->base_size, cudaMemAdviseSetAccessedBy, device));
+  dh::safe_cuda(
+      cudaMemPrefetchAsync(handle_->base_ptr, handle_->base_size, device, dh::DefaultStream()));
```

Comment:
Does this always happen after the data is initialized on the cpu?

Reply:
Yes, for now. We might optionally only prefetch the gradient index part and leave other scalars on host.